posthog-react-native/src/storage.ts

```diff
@@ -1,19 +1,75 @@
-import { PostHogCustomAsyncStorage } from './types'
+import { PostHogCustomAsyncStorage, PostHogCustomSyncStorage } from './types'
 
 const POSTHOG_STORAGE_KEY = '.posthog-rn.json'
 const POSTHOG_STORAGE_VERSION = 'v1'
 
 type PostHogStorageContents = { [key: string]: any }
 
+export class Storage {
+  memoryCache: PostHogStorageContents = {}
+  storage: PostHogCustomAsyncStorage | PostHogCustomSyncStorage
+  isPreloaded = false
+
+  constructor(storage: PostHogCustomAsyncStorage | PostHogCustomSyncStorage) {
+    this.storage = storage
+  }
+
+  persist(): void {
+    const payload = {
+      version: POSTHOG_STORAGE_VERSION,
+      content: this.memoryCache,
+    }
+
+    void this.storage.setItem(POSTHOG_STORAGE_KEY, JSON.stringify(payload))
+  }
+
+  getItem(key: string): any | null | undefined {
+    return this.memoryCache[key]
+  }
+  setItem(key: string, value: any): void {
+    this.memoryCache[key] = value
+    this.persist()
+  }
+  removeItem(key: string): void {
+    delete this.memoryCache[key]
+    this.persist()
+  }
+  clear(): void {
+    for (const key in this.memoryCache) {
+      delete this.memoryCache[key]
+    }
+    this.persist()
+  }
+  getAllKeys(): readonly string[] {
+    return Object.keys(this.memoryCache)
+  }
+
+  populateMemoryCache(res: string | null): void {
+    try {
+      const data = res ? JSON.parse(res).content : {}
+
+      for (const key in data) {
+        this.memoryCache[key] = data[key]
+      }
+    } catch (e) {
+      console.warn(
+        "PostHog failed to load persisted data from storage. This is likely because the storage format is. We'll reset the storage.",
```

Comment:
@benjackwhite I dont think we reset the storage though, the comment is outdated?